core/src/main/java/org/elasticsearch/indices/recovery/RecoverySource.java

```diff
@@ -144,38 +130,57 @@ public void messageReceived(final StartRecoveryRequest request, final TransportC
         }
     }
 
-    private static final class OngoingRecoveries {
-        private final Map<IndexShard, Set<RecoverySourceHandler>> ongoingRecoveries = new HashMap<>();
+    private final class OngoingRecoveries {
+        private final Map<IndexShard, ShardRecoveryContext> ongoingRecoveries = new HashMap<>();
+
+        synchronized RecoverySourceHandler add(StartRecoveryRequest request, IndexShard shard) {
+            final RecoverySourceHandler handler;
+            final ShardRecoveryContext currentContext = ongoingRecoveries.get(shard);
+            if (currentContext == null) {
+                final ShardRecoveryContext newContext = new ShardRecoveryContext();
+                handler = newContext.addNewRecovery(() -> createRecoverySourceHandler(request, shard, newContext));
+                ongoingRecoveries.put(shard, newContext);
+            } else {
+                handler = currentContext.addNewRecovery(() -> createRecoverySourceHandler(request, shard, currentContext));
+            }
+            return handler;
+        }
 
-        synchronized void add(IndexShard shard, RecoverySourceHandler handler) {
-            Set<RecoverySourceHandler> shardRecoveryHandlers = ongoingRecoveries.get(shard);
-            if (shardRecoveryHandlers == null) {
-                shardRecoveryHandlers = new HashSet<>();
-                ongoingRecoveries.put(shard, shardRecoveryHandlers);
+        private RecoverySourceHandler createRecoverySourceHandler(StartRecoveryRequest request, IndexShard shard,
```

Comment:
can we move this to ShardRecoveryContext and pass request and shard to addNewRecovery ? this way won't need all the supplier fun. Might as well rename the add method here to addNewRecovery too? or maybe `registerRecovery`?  (I don't mind much about the names, just a suggestion)